### PR TITLE
Convert from multi-index to storage index without canonicalization

### PR DESCRIPTION
**Pull request by macedo22**
## Proposed changes

This PR removes (what I believe to be) the unnecessary canonicalization of tensor multi-indices in converting them to their corresponding `Tensor` storage indices. The motivation for this is to improve the efficiency of `Structure::get_storage_index`.

`Structure` provides a means to convert between these through index mappings and functions that it defines. [Here](https://github.com/sxs-collaboration/spectre/blob/2693c35b13a4eeeb1a6f55847ea0ba98ba93987f/src/DataStructures/Tensor/Structure.hpp#L294) are the different types of index terms that it defines. The key parts of  `Structure` for this PR as I understand them are:
1. There is a 1:1 mapping between tensor multi-indices and collapsed indices. The collapsed indices appear to be just like a flattening of  a multi-dimensional array into a 1D array.
2. `Structure::collapsed_to_storage_` is a not-necessarily-1:1 mapping of collapsed indices to storage indices, depending on whether symmetries are present. It appears to be that, in generating this array, each unique multi-index is canonicalized ([short example comment of canonicalization](https://github.com/sxs-collaboration/spectre/blob/2693c35b13a4eeeb1a6f55847ea0ba98ba93987f/src/DataStructures/Tensor/Structure.hpp#L156)), and all multi-indices with the same canonical form are assigned the same storage index.
3. You can give the function, `Structure::get_storage_index`, a multi-index to get its corresponding storage index back out. Currently, this function first canonicalizes the given multi-index before computing its collapsed index, and then indexes into `collapsed_to_storage_` with the computed collapsed index to retrieve the storage index.

However, because there is a 1:1 mapping between multi-indices and collapsed indices, I don't believe this canonicalization step in `get_storage_index` is needed - I believe computing the collapsed index of the the original `tensor_index` argument in `get_storage_index` suffices since `collapsed_to_storage_` contains a storage index for each multi-index, not just the subset of them in the canonical form.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

**Review by nilsdeppe — DISMISSED** [review]
Yes, I agree.

I'm happy to merge as-is, but I'll ask you to consider adding some documentation to `compute_collapse_to_storage` (and some of the other similar functions?) to make it clear where a canonicalized index is necessary (it appears nowhere, right?)

**Comment by macedo22** [comment]
@nilsdeppe Sure! Would you like me to do this in a separate PR or as an additional commit here?

Yes, I don't think anywhere in `Structure.hpp` there are comments like "we canonicalize here because..." or "no need to canonicalize here because..." but [this](https://github.com/sxs-collaboration/spectre/blob/5234f36203a5c05837c146a449e1e52aa9716b9d/src/DataStructures/Tensor/Structure.hpp#L160) might be the only thing like that, as it states that we need to canonicalize to determine which multi-indices will have the same storage index, which are the ones with the same canonical form.

**Comment by nilsdeppe** [comment]
Up to you :) whichever is easier for you. I don't think adding it to this PR would slow things down

**Comment by macedo22** [comment]
@nilsdeppe Sounds good. I'll add some documentation in a new commit here and ping you after

**Comment by macedo22** [comment]
@nilsdeppe I added more documentation on the index mapping, overall. Some of it may be redundant, but I wasn't sure where to explain and where to skimp on explaining.

It seems that the documentation isn't generated for the contents of `Structure.hpp` currently on https://spectre-code.org/ or with my changes (locally with `make doc`). Am I perhaps just looking in the wrong places online and I'm missing the documentation somewhere?

**Review by nilsdeppe — DISMISSED** [review]
@fmahebert could you please take a look at this when you get some time? Thanks!

File: src/DataStructures/Tensor/Structure.hpp [hunk 1]
```diff
@@ -413,8 +470,17 @@ struct Structure {
...
 
-  /// Get storage_index
-  /// \param args comma separated list of the index to return
+  /// \brief Get the storage_index of a tensor_index
+  ///
+  /// \details
+  /// As there is a 1-1 mapping between a tensor_index and collapsed_index, and
+  /// collapsed_to_storage_ is a computed mapping between a collapsed_index to a
+  /// storage_index, we first compute the collapsed_index of the given
+  /// tensor_index and retrieve the `storage_index` from collapsed_to_storage_.
```
Comment (fmahebert) [inline src/DataStructures/Tensor/Structure.hpp]: [non-binding comment, up to you if you like this suggestion or no] I found this a bit cumbersome to read... would it still be accurate to write something shorter, perhaps like this?

First compute the collapsed_index of the given tensor_index (this is a 1-1 mapping), then retrieve the storage_index from collapsed_to_storage_.

(same in the two similar comments below)

File: src/DataStructures/Tensor/Structure.hpp [hunk 2]
```diff
@@ -391,15 +442,21 @@ struct Structure {
...
-   * `Symmetry<1, 1>`, this will return, e.g. `{{3, 2}}` rather than `{{2, 3}}`
-   * for that particular index.
-   * Note that this ordering is implementation defined.
-   */
+  /// \brief Get the canonical tensor_index array of a storage_index
+  ///
+  /// \details
+  /// For a symmetric tensor \f$T_{(ab)}\f$ with an associated symmetry list
+  /// `Symmetry<1, 1>`, this will return, e.g. `{{3, 2}}` rather than `{{2, 3}}`
+  /// for that particular index. Note that this ordering is implementation
```
Comment (fmahebert) [inline src/DataStructures/Tensor/Structure.hpp]: I suggest "this ordering" -> "the canonical ordering"

File: src/DataStructures/Tensor/Structure.hpp [hunk 3]
```diff
@@ -391,15 +442,21 @@ struct Structure {
...
+  /// defined.
+  ///
+  /// As `storage_to_tensor_` is a computed 1-1 mapping between a storage_index
+  /// and canonical tensor_index, we simply retrieve the canonical tesnor_index
```
Comment (fmahebert) [inline src/DataStructures/Tensor/Structure.hpp]: Typo: tesnor -> tensor

File: src/DataStructures/Tensor/Structure.hpp [hunk 4]
```diff
@@ -340,10 +382,19 @@ struct Structure {
...
+  /// correspond to tensor_indices with the same canonical form will map to the
+  /// same storage_index.
   static constexpr auto collapsed_to_storage_ =
       compute_collapsed_to_storage<Symm, number_of_components()>(
           make_cpp20_array_from_list<tmpl::conditional_t<
               sizeof...(Indices) == 0, size_t, index_list>>());
+  /// A 1-1 mapping between each storage_index and its canonical tensor_index.
+  /// If symmetries are present, each tensor_index is already in the canonical
+  /// form, and one that is unique, so this would equivalently mean a 1-1
+  /// mapping between each storage_index and tensor_index, for this case.
```
Comment (fmahebert) [inline src/DataStructures/Tensor/Structure.hpp]: I'm a little confused by this sentence (and the following "Otherwise..." sentence)... it seems to be saying the opposite from the docs of `compute_storage_to_tensor`... specifically you're making the same claim here "if symmetries are present" that is made in the docs above for the case "if no symmetries are present". Am I misreading, or is this written backwards?

[optional] Could also refer the reader to the docs for `compute_storage_to_tensor` to avoid having to restate things (and reduce chances of consistency mistakes :P)

File: src/DataStructures/Tensor/Structure.hpp [hunk 5]
```diff
@@ -299,8 +337,12 @@ struct ComponentNameImpl<0> {
 ///                     unique collapsed indices and there is a 1-1 map between
 ///                     a tensor_index and a collapsed_index.
 /// 3. storage_index: index into the storage vector of the Tensor. This depends
-///                   on symmetries of the tensor, rank and dimensionality.
-///                   There are size storage_index's.
+///                   on symmetries of the tensor, rank and dimensionality. When
+///                   symmetries are present in the tensor, while tensor_indices
+///                   that are equivalent due to symmetry will have different
+///                   collapsed_indices, they will have the same canonical form
+///                   and storage index. As such, there is a 1-1 map between
```
Comment (fmahebert) [inline src/DataStructures/Tensor/Structure.hpp]: [non-binding comment, up to you if you like this suggestion or no] The sentence "When symmetries..." was a bit hard to follow. I suggest rephrasing as something like,

If the Tensor has symmetries, tensor_indices that are equivalent due to symmetry will have the same canonical form, the same storage index, but different collapsed_indices.

File: src/DataStructures/Tensor/Structure.hpp [hunk 6]
```diff
@@ -340,10 +382,19 @@ struct Structure {
     return number_of_components;
   }
 
+  /// A mapping between each collapsed_index and its storage_index. If
+  /// symmetries are present, this will not be 1-1, as collapsed_indices that
+  /// correspond to tensor_indices with the same canonical form will map to the
+  /// same storage_index.
```
Comment (fmahebert) [inline src/DataStructures/Tensor/Structure.hpp]: [optional] Refer the reader to the docs for `compute_collapsed_to_storage` to avoid having to restate things here.

File: src/DataStructures/Tensor/Structure.hpp [hunk 7]
```diff
@@ -180,6 +197,27 @@ constexpr auto compute_collapsed_to_storage(
...
+/// \brief Computes a 1-1 mapping from a storage_index to its canonical
+/// tensor_index
+///
+/// \details
+/// When symmetries are present, not all unique tensor_indices can be retrieved
+/// from this map, as some tensor_indices will share the same canonical form.
+/// Otherwise, if no symmetries are present, each unique tensor_index is already
+/// in the canonical form, and one that is not shared by another tensor_index,
+/// so this would equivalently mean a 1-1 mapping from a storage_index to a
+/// tensor_index for each unique tensor_index. This means that when no
```
Comment (fmahebert) [inline src/DataStructures/Tensor/Structure.hpp]: I suggest removing the final "for each unique tensor_index" because I'm not sure what it means to have a 1-1 map from A to B for each A...

**Review by fmahebert — COMMENTED** [review]
Code changes look good, but I've a few suggestions to hopefully help clarity in the new docs. Thanks for writing these up, it's super helpful to have this spelled out clearly!

**Comment by fmahebert** [comment]
LGTM, please squash :)

**Comment by macedo22** [comment]
Squashed! Unfortunately, I'm still not sure how to get `Structure.hpp` documentation to appear. It seems that the documentation for neither `Structure` nor `TensorIndexType` appear on SpECTRE's site (as far as I can tell), and I see that what they have in common is that they are both in the `Tensor_detail` namespace, so I'm thinking it has something to do with that.

I'm currently troubleshooting this locally, but it takes some time to copy the generated documentation to my local machine, so I'm not sure how soon I'll have a fix ready. If either of you two have ideas as to why virtually none of the `Structure.hpp` documentation is generated, let me know. I'm not sure how to debug outside of trial and error with guessing and checking. All I have tried so far (locally, not pushed to a branch) is adding documentation for the `Tensor_detail` namespace in `Tensor.hpp` [here](https://github.com/sxs-collaboration/spectre/blob/088c3fff224bbcc11b6f580486aa750164fa1363/src/DataStructures/Tensor/Tensor.hpp#L42) like so, but it didn't seem to work:

```
/// \ingroup TensorGroup
/// Functions and classes that handle the lower level implementation details
/// of \ref Tensor "Tensor".
namespace Tensor_detail {
template <typename T, typename = void>
struct is_tensor : std::false_type {};
template <typename X, typename Symm, typename IndexList>
struct is_tensor<Tensor<X, Symm, IndexList>> : std::true_type {};
}  // namespace Tensor_detail
```

I thought this might kind of "mark" the `Tensor_detail` namespace to be something to generate documentation for. Next I'm going to attempt adding this documentation to above the `Tensor_detail` namespace in both `Structure.hpp` and `IndexType.hpp` instead.


**Update**: the second attempt mentioned just above also did not work. Will continue to try other ways to get `Tensor_detail` documentation to generate. Happy to hear suggestions!

**Comment by fmahebert** [comment]
@nilsdeppe probably knows, but I _think_ we don't generate Doxygen for anything in a "detail" or "blah_detail" namespace.

**Comment by nilsdeppe** [comment]
Yea, that's right

**Review by nilsdeppe — APPROVED** [review]
